lib/experimental/class-wp-html-updater.php

```diff
@@ -0,0 +1,556 @@
+				$this->set_attribute( 'class', $this->current_class_names_set() . '' );
+			} else {
+				$this->remove_attribute( 'class' );
+			}
+		}
+		$this->current_tag              = null;
+		$this->current_tag_name_ends_at = null;
+		$this->current_tag_attributes   = array();
+		$this->current_tag_class_names  = null;
+		$this->current_tag_touched_attributes_names       = array();
+	}
+
+	private function current_tag_matches( $tag_name, $class_name ) {
+		if ( $tag_name && ! WP_HTML_Comparator::equals( $this->current_tag, $tag_name ) ) {
+			// For debugging:
+			// echo "Tag name {$tag_name} does not match the current tag: {$this->current_tag}\n";
+			return false;
+		}
+		if ( $class_name ) {
+			$classes = $this->current_class_names_set();
+			// For debugging:
+			// echo "Class name {$class_name} does not match the current classes: {$classes}\n";
+			if ( ! $classes->has( $class_name ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	public function set_attribute( $name, $value ) {
+		if ( $this->current_tag ) {
+			$escaped_new_value = esc_attr( $value );
+			$updated_attribute = "{$name}=\"{$escaped_new_value}\"";
+
+			$attr = $this->find_current_tag_attribute( $name );
+			if ( $attr ) {
+				// Update an existing attribute.
+				$diff = new WP_HTML_Diff(
+					$attr->get_start_index(),
+					$attr->get_end_index(),
+					$updated_attribute
+				);
+			} else {
+				// Create a new attribute at the tag's name end.
+				$diff = new WP_HTML_Diff(
+					$this->current_tag_name_ends_at,
+					$this->current_tag_name_ends_at,
+					" {$updated_attribute}"
+				);
+			}
+			$this->add_diff( $name, $diff );
+		}
+
+		return $this;
+	}
+
+	public function remove_attribute( $name ) {
+		if ( $this->current_tag ) {
+			$attr = $this->find_current_tag_attribute( $name );
+			if ( $attr ) {
+				$diff = new WP_HTML_Diff(
+					$attr->get_start_index(),
+					$attr->get_end_index(),
+					''
+				);
+				$this->add_diff( $name, $diff );
+			}
+		}
+
+		return $this;
+	}
+
+	private function find_current_tag_attribute( $name ) {
+		if ( array_key_exists( $name, $this->current_tag_attributes ) ) {
+			return $this->current_tag_attributes[ $name ];
+		}
+		do {
+			$attr = $this->parse_next_attribute();
+		} while ( $attr && ! $attr->name_equals( $name ) );
+
+		return $attr;
+	}
+
+	public function add_class( $class_name ) {
+		if ( $this->current_tag ) {
+			$this->current_class_names_set()->add( $class_name );
+		}
+
+		return $this;
+	}
+
+	public function remove_class( $class_name ) {
+		if ( $this->current_tag ) {
+			$this->current_class_names_set()->remove( $class_name );
+		}
+
+		return $this;
+	}
+
+	private function add_diff( $attribute_name, $diff ) {
+		if ( in_array( $attribute_name, $this->current_tag_touched_attributes_names, true ) ) {
+			throw new WP_HTML_Attribute_Already_Modified( "Only one change per tag attribute is supported and the attribute '{$attribute_name}' was already changed on tag {$this->current_tag}." );
+		}
```

Comment:
While I considered adding this I wasn't yet decided because I thought it was the kind of situation that we may not need to guard against. There are cases where I think it could make sense that people take an "easy" way out of doing multiple conditional checks and instead simply adjust the same attribute multiple times in a row, expecting the last to win.

In #42507 I don't yet handle this and it's dependent on `usort`. Multiple changes stack up in the same location. I've been contemplating the value of the added complexity to either _prevent_ this situation, or more likely, maintain some consistent behavior when it happens.

Reply:
Good note – here's a different attempt: https://github.com/WordPress/gutenberg/pull/42485/commits/42b06a42e6391d0c020324b3f421ee9cf49e7437